site/_posts/2018-02-01-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+there was a leak in our `MetricsCollector` and `MetricsProcessor` workers\*.
+This was a trickle leak, where we would leak roughly 5-10 MB per hour, but this
+was unbound memory growth, so eventual this would cause the workers to exceed
+memory thresholds and restart.  This did require some extra setup to replicate
+as well, as we needed something to simulate adding jobs for these workers to
+collect/process metrics from so they had work to do.
+
+Since these are both "queue based" workers, this meant the "work" that they did
+was dependent on external factors, and wasn't always a linear amount of work
+passed to the queue, making the growth variable and not consistent.  On top of
+that, the jobs that caused the leak seemed to be at random, and instrumenting
+the code to determine the leak was hard to do in a generic fashion. That is not
+to say we didn't try wrap our workers with some common ruby profiling tools (we
+did... many times...), but the result usually led us to wonder if were actually
+dealing with the ["Observer effect"][7] and introducing new problems instead of
+"observing" existing ones.
+
+We eventually found that the `MIQ Server` (our worker monitor process) was
+leaking as well, though, it took much more time for this to be truly observable
+under normal conditions (like a month).  But, since it had a more predictable
+workflow, this was the better candidate to do some deeper process of
+elimination testing.  It took some time to get a scenario where this was
+leaking consistently, but once we did, it was simpler to do A/B testing with
+it.
+
+
+Step 1.5:  Reproducing the issue... fasterer
+--------------------------------------------
+
+One problem with using the `MiqServer` process as the target for observing the
+leak is it took _**days**_/_**weeks**_ to determine if it was leaking or just
+the process taking it's sweet time time to stabilize it's memory usage.  This
+led to very long gaps between making a change and having observable results to
+view from it.
+
+One of the least invasive things that we eventually did was decrease the server
+`:monitor_poll` and `:worker_monitor_frequency` (SPOILER! more on the second
+one in a bit) values from `"5.seconds"` and `"15.seconds"` respectively to
+`"1.second"` for them both.  This decreased the wait time between each
+iteration of it's normal work and, fortunately, increased the leak rate.
+
+We also increased the frequency of normal method calls by doing things like
+`40.times { do_some_work }` in places where we thought the leak might be, and
+if there was rapid growth, we knew where to probe deeper.
+
+These are things that happened throughout the process of discovering the cause
+of the leak, and didn't necessarily happen in between this step and the next
+(hence the "Step 1.5").  That said, steps like this to accelerate the feedback
+loop like this should be done as early into the memory leak debugging process
```

Comment:
Steps to accelerate the feedback loop, such as this, should be done as early into the memory leak debugging process....